Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SocketerNetworkingService.cs

```diff
@@ -0,0 +1,188 @@
+        [SerializeField]
+        protected TCPConnectionManager net = null;
+
+        /// <summary>
+        /// IPAddress for the user (HoloLens & HoloLens 2)
+        /// </summary>
+        [Tooltip("IPAddress for the user (HoloLens & HoloLens 2)")]
+        [SerializeField]
+        protected string userAddress = "127.0.0.1";
+
+        /// <summary>
+        /// Port for the user (HoloLens & HoloLens 2)
+        /// </summary>
+        [Tooltip("Port for the user (HoloLens & HoloLens 2)")]
+        [SerializeField]
+        protected int userPort = 7997;
+
+        /// <summary>
+        /// Timeout interval for establishing connections
+        /// </summary>
+        [Tooltip("Timeout interval for establishing connections")]
+        [SerializeField]
+        protected TimeSpan timeoutInterval = TimeSpan.Zero;
+
+        private bool runAsUser = false;
+        private Dictionary<string, string> playerIds = new Dictionary<string, string>();
+
+        /// <inheritdoc />
+        public event PlayerConnectedHandler PlayerConnected;
+
+        /// <inheritdoc />
+        public event PlayerDisconnectedHandler PlayerDisconnected;
+
+        /// <inheritdoc />
+        public event DataHandler DataReceived;
+
+        private void Start()
+        {
+#if UNITY_WSA && !UNITY_EDITOR
+            runAsUser = true;
+#elif UNITY_ANDROID || UNITY_IOS || UNITY_EDITOR
+            runAsUser = false;
+#endif
+        }
+
+        /// <inheritdoc />
+        public void Connect()
+        {
+            if (net == null)
+            {
+                Debug.LogError("TCPConnectionManager not set for SocketNetworkingService. Unable to establish connection");
+                return;
+            }
+            else
+            {
+                net.DisconnectAll();
+            }
+
+            net.OnConnected += OnConnected;
+            net.OnDisconnected += OnDisconnected;
+            net.OnReceive += OnReceive;
+
+            if (runAsUser)
+            {
+                net.StartListening(userPort);
+            }
+            else
+            {
+                net.ConnectTo(userAddress, userPort);
+            }
+        }
+
+        private void OnConnected(SocketEndpoint obj)
+        {
+            string playerId = string.Empty;
+            lock(playerIds)
+            {
+                if (!playerIds.ContainsKey(obj.Address))
+                {
+                    playerIds.Add(obj.Address, Guid.NewGuid().ToString());
+                }
+
+                playerId = playerIds[obj.Address];
+            }
+
+            if (playerId != string.Empty)
+            {
+                PlayerConnected?.Invoke(playerId);
+            }
+        }
+
+        private void OnDisconnected(SocketEndpoint obj)
+        {
+            string playerId = string.Empty;
+            lock (playerIds)
+            {
+                if (playerIds.ContainsKey(obj.Address))
+                {
+                    playerId = playerIds[obj.Address];
```

Comment:
>playerId = playerIds[obj.Address]; [](start = 20, length = 34)

Should this also be getting removed from the dictionary at this point? Or do you want to be storing them forever for other reasons (i.e. still want to handle OnRecieve after disconnection?)

Reply:
In some instances the same address will reconnect and send payloads before the connection event occurs. To avoid having these messages dropped, I intentionally kept id definitions for the same address.

Reply:
I'm going to keep this for OnReceive after disconnection. If we unsubscribe we hopefully won't hit this issue, but i have seen some stranger behavior here where messages get received before a reconnect if the client/server disconnects.